core/utils/compare.go

```diff
@@ -0,0 +1,9 @@
+package utils
+
+func IsZero[C comparable](val C) bool {
+	var zero C
```

Comment:
Does this work always for all chains?
What if a chain has a 0 value that's not same as "var zero C"?

Reply:
this works for any type if it is follows the `comparable` constraint

Reply:
Nonzero address is certainly possible.
In Eth too, string representation of Zero-addres is "0x00000000".
Some chains could have a non-zero prefix to use for all addresses, including zero-prefix.
I vaguely remember Tron was doing something of sorts.

But it is fine for now. Ok to proceed with this, and we'll see when a different chain is encountered.

Reply:
https://go.dev/ref/spec#The_zero_value
> [_"Make the zero value useful."_](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=6m25s)

If a chain defines zero in a way that can't be represented as an uninitialized value, then we'll just have to redefine this abstraction to be more general :shrug: 

Hopefully we don't end up having to support IEE 754 weirdness, but we can if we need to: 
- https://en.wikipedia.org/wiki/IEEE_754#Special_values 
- https://en.wikipedia.org/wiki/Signed_zero

Reply:
yeah we can handle it if needed - maybe just a `Zero()` method associated with the type to provide the zero value rather than comparing it to an uninitialized value.

but i wouldn't worry about it until we come across it